### DISCUSSION TITLE
Customize the Axis labels with SVG/HTML

### DISCUSSION THREAD

**Discussion by Marttyn** [observablehq/plot#2190] [Q&A] [answered]
Hi,
I'm trying to create a bar chart similar to the one in the image below:

![image](https://github.com/user-attachments/assets/859226e6-fd6f-41f2-bd1a-ccfca521eb43)

I want to add country flags alongside the y-axis labels, but I’ve run into a couple of issues. I’m using React, and I have access to a library of flags in SVG format. However, it seems like the y-axis labels in the data array only accept strings, not SVG elements or React components.
![image](https://github.com/user-attachments/assets/f041ca9e-7f1b-4be7-a1d9-f1700f37bee7)

I’ve also tried using emoji flags (e.g., 🇧🇷 using this method: [Emojipedia](https://emojipedia.org/flag-brazil#technical)), but the flags don’t render correctly.
```
function getFlagEmoji(countryCode: string) {
  let codePoints = countryCode
    .toUpperCase()
    .split("")
    .map((char) => 127397 + char.charCodeAt(0));
  return String.fromCodePoint(...codePoints);
}
```
![image](https://github.com/user-attachments/assets/af884f30-508f-4dde-acd2-a8d1a1e55c70)

What would be the best way to include SVG flags in the y-axis labels? Is there a workaround for this limitation?

Thanks in advance!

**Comment by Fil** [comment, marked as answer]
Using emojis should work:

```
const codes = new Map([
  ["Argentina", "AR"],
  ["Bolivia", "BO"],
  ["Chile", "CL"],
  ["Brazil", "BR"]
]);

function getFlagEmoji(c) {
  const countryCode = codes.get(c) ?? c;
  let codePoints = countryCode
    .toUpperCase()
    .split("")
    .map((char) => 127397 + char.charCodeAt(0));
  return String.fromCodePoint(...codePoints);
}

Plot.barX(
  [
    [1, "Argentina"],
    [2, "Brazil"],
    [3, "Chile"],
    [3, "Bolivia"]
  ],
  { x: "0", y: "1" }
).plot({ marginLeft: 100, y: { tickFormat: (d) => `${getFlagEmoji(d)} ${d}` } })
```

But I agree it could be nice to accept a SVG node as the output of the tickFormat function.

**Reply by Marttyn** [reply]
Unfortunately, it doesn't, with emoji I get the following

```
Plot.axisY({
            label: null,
            tickSize: 0,
            fill: "#252931",
            fontFamily: "roboto condensed",
            fontSize: 16,
            fontWeight: 700,
            lineAnchor: "middle",
            textAnchor: "start",
            dx: -countryTextWidth,
            tickFormat: (d) => `${getFlagEmoji(d)} ${d}`,
          }),
```

![image](https://github.com/user-attachments/assets/b6e48e53-e328-4146-99f0-de6cdf05fec6)

**Reply by Fil** [reply]
Does it work here?
https://observablehq.com/@fil/emoji-flags-plot

this is what I'm seeing:
![Capture d’écran 2024-10-08 à 16 08 40](https://github.com/user-attachments/assets/c7fd89fa-ea28-4861-aa9f-313bbe324a6e)

**Reply by mbostock** [reply]
It depends on whether your chosen font supports color emoji.

**Reply by Marttyn** [reply]
For me it doesn't work, I tried in Chrome and Edge

![image](https://github.com/user-attachments/assets/cdcbf120-c049-4043-b387-46bde79a2b4a)

EDIT: I opened on my phone and it did work. I don't know why this difference.

**Reply by Fil** [reply]
mhhh ok, maybe windows vs linux vs macos? (I've tested on macos only)

**Reply by Marttyn** [reply]
You are right, I tested on Mac and Android and it did work.
Looks like Windows doesn't support flag emoji.
https://answers.microsoft.com/en-us/windows/forum/all/country-emoji-for-windows/78a3e92b-02a8-4e7c-814e-c1c0596eaf4e

So unfortunately, flag emoji is not a good solution for me, as I need Windows support too.

**Reply by Fil** [reply]
Maybe try the [image mark](https://observablehq.com/plot/marks/image)? Once you know the list of countries, you can use Plot.image() with **frameAnchor**: "left", **y** being the country name, and **src** being the URI of the flag svg (can be a data-uri).

Also, I think we should create a general feature request concerning both the **text** and the **tip** marks. (The axis marks use the text mark internally to render the ticks labels). Because we often want to extend these marks to render as custom svg or even HTML nodes. Ref. https://github.com/observablehq/plot/issues/1767 for the **tip** use case.

**Reply by Marttyn** [reply]
I found a solution for the emoji flags here: https://stackoverflow.com/questions/62729729/how-do-i-view-country-flags-on-windows-10-through-html

Using the font [Noto Emoji](https://github.com/googlefonts/noto-emoji) works on windows too.
or from Google: https://fonts.google.com/noto/specimen/Noto+Color+Emoji
